RFC: Avoid over-pessimization in apply_type_tfunc

martinholters opened the pull request:
Ref. https://github.com/JuliaLang/julia/pull/25861#issuecomment-387415946.

martinholters committed:
Avoid over-pessimization in apply_type_tfunc

martinholters commented:
Hm, LibGit2/libgit2 segfaults on 32bit Circle and Travis and hangs on FreeBSD), so likely related. Presumable, I now produce a `Union{}` were I shouldn't. Will try to reproduce locally. Or is that a failure we've seen elsewhere recently?

martinholters commented:
> Will try to reproduce locally.

Works locally with a 32bit build. Any ideas?

vtjnash commented on base/compiler/tfuncs.jl:
```diff
@@ -724,20 +724,30 @@ function apply_type_tfunc(@nospecialize(headtypetype), @nospecialize args...)
+            else
                 if !isType(ai)
-                    return Any
+                    if !isa(ai, Type) || ai <: Type || Type <: ai
```
typeintersect(ai, Type) != Union{}

martinholters replied:
Good catch!

vtjnash commented on base/compiler/tfuncs.jl:
```diff
@@ -724,20 +724,30 @@ function apply_type_tfunc(@nospecialize(headtypetype), @nospecialize args...)
+                if !isa(ai.val, Type)
+                    return Union{}
+                end
+            else
```
Checking explicitly for if isa TypeVar might be necessary (And PartialTypeVar)

vtjnash replied:
yes. or `ai` could be a PartialTypeVar

martinholters replied:
`ai` being a `PartialTypeVar` should be handled below. Do you mean `ai.val`?

Sacha0 commented:
> Hm, LibGit2/libgit2 segfaults on 32bit Circle and Travis and hangs on FreeBSD), so likely related. Presumable, I now produce a Union{} were I shouldn't. Will try to reproduce locally. Or is that a failure we've seen elsewhere recently?

Though slightly different, LibGit2 failures recently popped up in https://github.com/JuliaLang/julia/pull/26997 as well. Best!

martinholters commented on base/compiler/tfuncs.jl:
```diff
@@ -724,20 +724,34 @@ function apply_type_tfunc(@nospecialize(headtypetype), @nospecialize args...)
+            if isa(ai, Const)
+                if !isa(ai.val, Type)
+                    if isa(ai.val, TypeVar) || isa(ai.val, PartialTypeVar)
+                        return hasnonType = true
```
Did you mean like so, @vtjnash?

martinholters replied:
Bump. I'm still not sure you meant this. A `Const` `PartialTypeVar` looks a bit funny here to me.

vtjnash replied:
`isa(ai.val, PartialTypeVar)` can return `Union{}`. But `ai` itself can be a `PartialTypeVar`

martinholters replied:
Alright, then I've apparently misinterpreted your earlier remark. The `isa(ai, PartialTypeVar)` case is already handled below (in the `!isType(ai)` and `!isa(ai, Type)` branch).

martinholters commented:
Oh man, CI really doesn't like this one. AV timed out, an occasional libgit2 failure, end the rest
```
Worker 3 failed running test OldPkg/pkg:
Some tests did not pass: 2 passed, 0 failed, 1 errored, 0 broken.OldPkg/pkg: Error During Test at /tmp/julia/share/julia/test/testdefs.jl:19
  Got exception LoadError("/tmp/julia/share/julia/stdlib/v0.7/OldPkg/test/pkg.jl", 56, KeyError("SparseArrays")) outside of a @test
  LoadError: KeyError: key "SparseArrays" not found
```
or `s/SparseArrays/Printf`. All of these are commonly seen, and tests get pretty far before hitting them, so this is probably ok. OTOH, this is not urgent at all, so we can just rerun CI once it feels better. Meanwhile, @vtjnash, could you verify whether I've (correctly) addressed all your comments?

martinholters commented:
Oh, rebasing this led to an all-green CI! As a consequence, I'm unsure whether I'll ever dare touching this again. So...good to go?

StefanKarpinski commented:
I would give one of @JeffBezanson, @Keno or @vtjnash 24 hours to comment and then merge :trollface:

StefanKarpinski commented:
Ooo! Requesting a review from @nanosoldier, fancy.

martinholters commented:
Fancy indeed. But apparently, doesn't trigger him, so @nanosoldier `runbenchmarks(ALL, vs=":master")`.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/27150#issuecomment-396483499) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/6dae1c12e6e13d3ae135d20af4f5e39f681f0f2e/737ac67_vs_83ce7ba/report.md). cc @ararslan

martinholters commented:
I'm inclined to declare regressions as well as improvements there noise and will merge soonish unless anyone objects.
